src/pudl/analysis/demand_mapping.py

```diff
@@ -1,294 +1,1009 @@
+        attributes dictionary
     """
-    intersection_matrix = gdf_intersection.pivot_table(
-        values=gdf_intersection_col,
-        index=gdf_intermediate_col,
-        columns=gdf_source_col,
-        fill_value=0
-    )
+    # ID is the index which will help to identify duplicate geometries
+    gdf_ids = [k for k, v in attributes.items() if (
+        (k in gdf.columns) and (v == "ID"))][0]
+    gdf_constants = [k for k, v in attributes.items() if (
+        (k in gdf.columns) and (v == "constant"))]
+    gdf_uniforms = [k for k, v in attributes.items() if (
+        (k in gdf.columns) and (v == "uniform"))]
+
+    # Check if `ID` column has all unique elements:
+    if gdf[gdf_ids].nunique() != len(gdf):
+        raise Exception("All ID column elements should be unique")
+
+    # Iterating through each of the geometries
+    for row in tqdm((gdf[[gdf_ids, "geometry"]]
+                     .reset_index(drop=True)
+                     .itertuples()), total=len(gdf)):
+
+        index = row.Index
+        if index == 0:
+            gdf_disjoint = (pd.DataFrame(row).T
+                            .drop(0, axis=1)
+                            .rename(columns={1: gdf_ids, 2: "geometry"}))
+            gdf_disjoint[gdf_ids] = gdf_disjoint[gdf_ids].apply(
+                lambda x: frozenset([x]))
+            gdf_disjoint = GeoDataFrame(
+                gdf_disjoint, geometry="geometry", crs=gdf.crs)
+            gdf_disjoint_cur_union = unary_union(gdf_disjoint["geometry"])
+
+        # Additional geometries
+        else:
+
+            # Adding difference and intersections of the old geometries
+            # with the new geometry
+            gdf_disjoint["geometry_new_diff"] = gdf_disjoint.difference(
+                row[2])
+            gdf_disjoint["geometry_new_int"] = gdf_disjoint.intersection(
+                row[2])
+            gdf_disjoint = gdf_disjoint.drop("geometry", axis=1)
+
+            # Stacking all the new geometries in one column
+            gdf_disjoint = (gdf_disjoint
+                            .set_index(gdf_ids)
+                            .stack()
+                            .reset_index()
+                            .rename(columns={"level_1": "geom_type", 0: "geometry"}))
+
+            # Creating the new gdf_ids sets
+            gdf_disjoint[gdf_ids] = gdf_disjoint.apply(
+                lambda x: edit_id_set(x, row[1], gdf_ids), axis=1)
+
+            # Adding the new sole geometry's gdf_ids and geometry
+            gdf_disjoint = gdf_disjoint.append({
+                gdf_ids: frozenset([row[1]]),
+                "geom_type": "geometry_new_sole",
+                "geometry": row[2].difference(gdf_disjoint_cur_union)
+            }, ignore_index=True)
+
+            # Removing geometries which are not polygons
+            gdf_disjoint["geometry"] = gdf_disjoint["geometry"].apply(
+                lambda x: polygonize_geom(x))
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Lambda may not be necessary](https://app.codacy.com/manual/zaneselvans/pudl/pullRequest?prid=5664147)